packages/liveblocks-react-tiptap/src/toolbar/shared.tsx

```diff
@@ -0,0 +1,57 @@
+import { useLayoutEffect } from "@liveblocks/react/_private";
+import {
+  Children,
+  type ComponentProps,
+  createContext,
+  forwardRef,
+  useContext,
+  useId,
+  useMemo,
+} from "react";
+
+// This file is separate to avoid circular dependencies
+
+type FloatingToolbarContext = {
+  close: () => void;
+  registerExternal: (id: string) => () => void;
+};
+
+export const FloatingToolbarContext =
+  createContext<FloatingToolbarContext | null>(null);
+
+export const FloatingToolbarExternal = forwardRef<
+  HTMLDivElement,
+  ComponentProps<"div">
+>(({ children, style, ...props }, forwardedRef) => {
+  const id = useId();
+  const externalId = useMemo(
+    () => `liveblocks-floating-toolbar-external-${id}`,
+    [id]
+  );
```

Comment:
This really isn't necessary for simple computations.

```suggestion
  const externalId = `liveblocks-floating-toolbar-external-${id}`;
```